### ISSUE TITLE
RoR 4.2.4: Custom fonts not loaded in production

### ISSUE THREAD

**Issue by Sega100500** [rails/rails#21367]
1. In config/environments/production.rb
config.assets.compile = true

2. fonts directory structure:
app/assets/stylesheets/_fonts.scss:
@import "font/RobotoBold.css";

app/assets/stylesheets/font/RobotoBold.css:
@font-face {
    font-family: 'RobotoBold';
    src: url('Roboto/RobotoBold.eot');
    src: url('Roboto/RobotoBold.eot') format('embedded-opentype'),
         url('Roboto/RobotoBold.woff2') format('woff2'),
         url('Roboto/RobotoBold.woff') format('woff'),
         url('Roboto/RobotoBold.ttf') format('truetype'),
         url('Roboto/RobotoBold.svg#RobotoBold') format('svg');
}

in app/assets/stylesheets/font/Roboto directory - files of font

3. Attach fonts
@import "fonts";

In RoR 4.2.3 ALL WORK FINE! In 4.2.4 fonts not loaded, displays as standart fonts.
In development mode all work.

Please fix it! Many project now displayed incorrectly

**Comment by rafaelfranca** [comment]
I don't see how this was working in any version of Rails 4 because the fonts urls are not using the assets helpers. Could you change your css to use the assets helper and report back if is still an issue? Thanks

**Comment by Sega100500** [comment]
@rafaelfranca 
You are idiot? Many sorry!...
I SAY - In RoR 4.2.3 ALL WORK FINE!
I just updated to 4.2.4 and fonts are not displayed
In ALL versions (include ALL Rails 4 before 4.2.4) it WORK!

**Comment by Sega100500** [comment]
@rafaelfranca 
In config/initializers/assets.rb
Rails.application.config.assets.precompile += ['*.js', '*.css', '**/*.js', '**/*.css']

Why I need to REWRITE MANY .css files (of fonts) generated by generator to use in RoR 4.2.4. Just because someone turned curves hands.

**Comment by rafaelfranca** [comment]
Sorry, Rails 4.2.4 didn't changed anything about the assets pipeline, in fact Rails is not even responsible for that, so I'm pretty sure it was not the Rails release. Like I said, the code you posted is not supposed to work because your css has `url` calls pointing to non digested assets. That is not supported since sprockets-rails 2 that was released with Rails 4. If you want to check that create a new Rails application using Rails 4.2.3 and try to get it working.

I'm sorry, I didn't wanted to upset you, but I'm really confident that it is not a rails bug.

**Comment by Sega100500** [comment]
@rafaelfranca 
I have many projects! Rails 3, Rails 4 - ALL works!
I just updated to 4.2.4 and it NOT WORKING

Miracles?

**Comment by rafaelfranca** [comment]
Could you post the gemfile.lock diff when you updated to Rails 4.2.4? It may be related to other gem

**Comment by rafaelfranca** [comment]
(the gemfile content would help too)

**Comment by Sega100500** [comment]
@rafaelfranca 
Gem.lock
GEM
  remote: https://rubygems.org/
  specs:
    actionmailer (4.2.4)
      actionpack (= 4.2.4)
      actionview (= 4.2.4)
      activejob (= 4.2.4)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 1.0, >= 1.0.5)
    actionpack (4.2.4)
      actionview (= 4.2.4)
      activesupport (= 4.2.4)
      rack (~> 1.6)
      rack-test (~> 0.6.2)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (4.2.4)
      activesupport (= 4.2.4)
      builder (~> 3.1)
      erubis (~> 2.7.0)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    activejob (4.2.4)
      activesupport (= 4.2.4)
      globalid (>= 0.3.0)
    activemodel (4.2.4)
      activesupport (= 4.2.4)
      builder (~> 3.1)
    activerecord (4.2.4)
      activemodel (= 4.2.4)
      activesupport (= 4.2.4)
      arel (~> 6.0)
    activesupport (4.2.4)
      i18n (~> 0.7)
      json (~> 1.7, >= 1.7.7)
      minitest (~> 5.1)
      thread_safe (~> 0.3, >= 0.3.4)
      tzinfo (~> 1.1)
    arel (6.0.3)
    bcrypt (3.1.10)
    binding_of_caller (0.7.2)
      debug_inspector (>= 0.0.1)
    builder (3.2.2)
    byebug (6.0.2)
    ckeditor (4.1.1)
      cocaine
      orm_adapter (~> 0.5.0)
    climate_control (0.0.3)
      activesupport (>= 3.0)
    cocaine (0.5.7)
      climate_control (>= 0.0.3, < 1.0)
    coffee-rails (4.1.0)
      coffee-script (>= 2.2.0)
      railties (>= 4.0.0, < 5.0)
    coffee-script (2.4.1)
      coffee-script-source
      execjs
    coffee-script-source (1.9.1.1)
    debug_inspector (0.0.2)
    erubis (2.7.0)
    execjs (2.6.0)
    globalid (0.3.6)
      activesupport (>= 4.1.0)
    i18n (0.7.0)
    jbuilder (2.3.1)
      activesupport (>= 3.0.0, < 5)
      multi_json (~> 1.2)
    jquery-rails (4.0.4)
      rails-dom-testing (~> 1.0)
      railties (>= 4.2.0)
      thor (>= 0.14, < 2.0)
    json (1.8.3)
    libv8 (3.16.14.11)
    loofah (2.0.3)
      nokogiri (>= 1.5.9)
    mail (2.6.3)
      mime-types (>= 1.16, < 3)
    mime-types (2.6.1)
    mimemagic (0.3.0)
    mini_magick (4.2.10)
    mini_portile (0.6.2)
    minitest (5.8.0)
    multi_json (1.11.2)
    mysql2 (0.3.19)
    nokogiri (1.6.6.2)
      mini_portile (~> 0.6.0)
    orm_adapter (0.5.0)
    paperclip (4.3.0)
      activemodel (>= 3.2.0)
      activesupport (>= 3.2.0)
      cocaine (~> 0.5.5)
      mime-types
      mimemagic (= 0.3.0)
    passenger (5.0.16)
      rack
      rake (>= 0.8.1)
    rack (1.6.4)
    rack-test (0.6.3)
      rack (>= 1.0)
    rails (4.2.4)
      actionmailer (= 4.2.4)
      actionpack (= 4.2.4)
      actionview (= 4.2.4)
      activejob (= 4.2.4)
      activemodel (= 4.2.4)
      activerecord (= 4.2.4)
      activesupport (= 4.2.4)
      bundler (>= 1.3.0, < 2.0)
      railties (= 4.2.4)
      sprockets-rails
    rails-deprecated_sanitizer (1.0.3)
      activesupport (>= 4.2.0.alpha)
    rails-dom-testing (1.0.7)
      activesupport (>= 4.2.0.beta, < 5.0)
      nokogiri (~> 1.6.0)
      rails-deprecated_sanitizer (>= 1.0.1)
    rails-html-sanitizer (1.0.2)
      loofah (~> 2.0)
    railties (4.2.4)
      actionpack (= 4.2.4)
      activesupport (= 4.2.4)
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (10.4.2)
    rdoc (4.2.0)
    ref (2.0.0)
    responders (2.1.0)
      railties (>= 4.2.0, < 5)
    sass (3.4.17)
    sass-rails (5.0.3)
      railties (>= 4.0.0, < 5.0)
      sass (~> 3.1)
      sprockets (>= 2.8, < 4.0)
      sprockets-rails (>= 2.0, < 4.0)
      tilt (~> 1.1)
    sdoc (0.4.1)
      json (~> 1.7, >= 1.7.7)
      rdoc (~> 4.0)
    sitemap_generator (5.1.0)
      builder
    sprockets (3.3.3)
      rack (~> 1.0)
    sprockets-rails (2.3.2)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      sprockets (>= 2.8, < 4.0)
    therubyracer (0.12.2)
      libv8 (~> 3.16.14.0)
      ref
    thor (0.19.1)
    thread_safe (0.3.5)
    tilt (1.4.1)
    tzinfo (1.2.2)
      thread_safe (~> 0.1)
    uglifier (2.7.1)
      execjs (>= 0.3.0)
      json (>= 1.8.0)
    web-console (2.2.1)
      activemodel (>= 4.0)
      binding_of_caller (>= 0.7.2)
      railties (>= 4.0)
      sprockets-rails (>= 2.0, < 4.0)
    will_paginate (3.0.7)

PLATFORMS
  ruby

DEPENDENCIES
  admin!
  bcrypt (~> 3.1.7)
  byebug
  ckeditor (= 4.1.1)
  coffee-rails (~> 4.1.0)
  jbuilder (~> 2.0)
  jquery-rails
  mini_magick
  mysql2
  paperclip
  passenger
  rails (~> 4.2.0)
  responders (~> 2.0)
  sass-rails (~> 5.0)
  sdoc (~> 0.4.0)
  sitemap_generator
  therubyracer
  uglifier (>= 1.3.0)
  web-console (~> 2.0)
  will_paginate

BUNDLED WITH
   1.10.6

**Comment by Sega100500** [comment]
@rafaelfranca 
source 'https://rubygems.org'


# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
gem 'rails', '~> 4.2.0'
# Use mysql as the database for Active Record
gem 'mysql2'
# Use SCSS for stylesheets
gem 'sass-rails', '~> 5.0'
# Use Uglifier as compressor for JavaScript assets
gem 'uglifier', '>= 1.3.0'
# Use CoffeeScript for .coffee assets and views
gem 'coffee-rails', '~> 4.1.0'
# See https://github.com/sstephenson/execjs#readme for more supported runtimes
gem 'therubyracer', platforms: :ruby

# Use jquery as the JavaScript library
gem 'jquery-rails'
# Turbolinks makes following links in your web application faster. Read more: https://github.com/rails/turbolinks
###gem 'turbolinks'
# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
gem 'jbuilder', '~> 2.0'
# bundle exec rake doc:rails generates the API under doc/api.
gem 'sdoc', '~> 0.4.0', group: :doc

# Use ActiveModel has_secure_password
gem 'bcrypt', '~> 3.1.7'

# Use Unicorn as the app server
# gem 'unicorn'

# Use Capistrano for deployment
# gem 'capistrano-rails', group: :development

group :development, :test do
  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
  gem 'byebug'

  # Access an IRB console on exception pages or by using <%= console %> in views
  gem 'web-console', '~> 2.0'

  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
  ### gem 'spring'
end


# +++

# Administration tools
#
# to install migrations: rake admin:install:migrations
gem 'admin', :path => 'admin'

# Paginate
gem 'will_paginate'

# CKEditor
#
# for install:
# rails generate ckeditor:install --orm=active_record --backend=paperclip
# rake db:migrate
#
gem 'ckeditor', '4.1.1'
gem 'paperclip'
gem 'responders', '~> 2.0'

# Mini_Magick
gem 'mini_magick'

# Passenger
#
# download: git clone https://github.com/FooBarWidget/passenger.git
# and: ./bin/passenger-install-apache2-module
#
# to install from gem: passenger-install-apache2-module
# stat: passenger-memory-stats

gem 'passenger'

# SitemapGenerator
gem 'sitemap_generator'

**Comment by rafaelfranca** [comment]
Could you confirm that sprockets upgraded from 2 to 3.3?

**Comment by Sega100500** [comment]
@rafaelfranca 
In Gemfile.lock
sprockets (3.3.3)

**Comment by rafaelfranca** [comment]
It was 2.x.x before the Rails upgrade?

**Comment by Sega100500** [comment]
@rafaelfranca 
No!
It along time ago upgraded
in 4.2.3 already sprockets 3.3.3

**Comment by Sega100500** [comment]
@rafaelfranca 
Ok!
I do not want to argue.
How do I change the url in .css fonts files for everything to work properly?

**Comment by rafaelfranca** [comment]
Right! Thanks for all information, I'll try to reproduce in a new application later

**Comment by Sega100500** [comment]
@rafaelfranca 
Now I need to all works
What I do change in fonts css files? url?

**Comment by rafaelfranca** [comment]
I create a test application but could not make it fail https://github.com/rafaelfranca/fonts_test. Could you check it and change what is required to make it fail?

**Comment by Sega100500** [comment]
@rafaelfranca 
In production mode it work ? Correct display Roboto font ?

**Comment by rafaelfranca** [comment]
Yes. You can try yourself

**Comment by Sega100500** [comment]
@rafaelfranca 
without 'rake assets:precompile' ?
with 'rake assets:precompile' it work for me, but I need WITHOUT it.

**Comment by rafaelfranca** [comment]
I'll try that too

**Comment by rafaelfranca** [comment]
Updated the repository and now I can reproduce even with Rails 4.2.3, can you check if that is the behavior that you are seeing?

**Comment by Sega100500** [comment]
@rafaelfranca 
Very strange! I like the same thing, but it is not working.
Do not remove until https://github.com/rafaelfranca/fonts_test. . I'll check what's wrong in my app.

**Comment by Sega100500** [comment]
@@rafaelfranca 
I create new RoR 4.2.4 app
Result - font not loaded in production

application.css:

/*
 * This is a manifest file that'll be compiled into application.css, which will include all the files
 * listed below.
 *
 * Any CSS and SCSS file within this directory, lib/assets/stylesheets, vendor/assets/stylesheets,
 * or any plugin's vendor/assets/stylesheets directory can be referenced here using a relative path.
 *
 * You're free to add application-wide styles to this file and they'll appear at the bottom of the
 * compiled file so the styles you add here take precedence over styles defined in any styles
 * defined in the other CSS/SCSS files in this directory. It is generally better to create a new
 * file per style scope.
 *
 *##=## require_tree .
 *= require_self
 *= require app
 */


app.scss:

@import "fonts";

body {
  font-family: $font_RobotoBold;
}


_fonts.scss:

@import "font/RobotoBold.css";

$font_RobotoBold: RobotoBold, Arial, Helvetica, sans-serif;

**Comment by Sega100500** [comment]
@rafaelfranca 
http://ft.inet777.pro/

**Comment by Sega100500** [comment]
After 'rake assets:precompile' all work. But! In config/initializers/assets.rbconfig/initializers/assets.rb I wrote config.assets.compile = true
It may still be a problem that can not be found, these files?

**Comment by Sega100500** [comment]
HOW I can made SAME behavior of application in development and production mode WITHOUT run 'rake assets:precompile' every time?

**Comment by Sega100500** [comment]
I find difference!
When I wrote in _fonts.scss:
@import "font/RobotoBold.css";
it work only in development mode with font/RobotoBold.css:
```
@font-face {
    font-family: 'RobotoBold';
    src: url('Roboto/RobotoBold.eot');
    src: url('Roboto/RobotoBold.eot') format('embedded-opentype'),
         url('Roboto/RobotoBold.woff2') format('woff2'),
         url('Roboto/RobotoBold.woff') format('woff'),
         url('Roboto/RobotoBold.ttf') format('truetype'),
         url('Roboto/RobotoBold.svg#RobotoBold') format('svg');
}
```

When I wrote @import "font/RobotoBold"; // WITHOUT extension
it work in development and production mode, BUT with font/RobotoBold.css:
```
@font-face {
    font-family: 'RobotoBold';
    src: url('font/Roboto/RobotoBold.eot');
    src: url('font/Roboto/RobotoBold.eot') format('embedded-opentype'),
         url('font/Roboto/RobotoBold.woff2') format('woff2'),
         url('font/Roboto/RobotoBold.woff') format('woff'),
         url('font/Roboto/RobotoBold.ttf') format('truetype'),
         url('font/Roboto/RobotoBold.svg#RobotoBold') format('svg');
}
```

It is with in assets.rb initializer:
```
Rails.application.config.assets.precompile += ['*.js', '*.css', '**/*.js', '**/*.css']
```

Very, VERY STRANGE behavior!
Perhaps, after all, this is a BUG?

**Comment by Sega100500** [comment]
Two different directives:
@import in scss
and
@import in css

When I write @import in css file, it work only for development mode properly.
@import "file_name" in scss is only "_file_name.scss" must be.

Maybe I'm wrong, but directive in css-file @import "filename.css" must be acessible to /assets/filename.css in production mode.

**Comment by mkhairi** [comment]
@Sega100500  it seems like sass v3.4.17 bug. i having the same issue. rollback to 3.4.16 will fix this.

**Comment by Sega100500** [comment]
When I run 'rake assets:precompile' all work fine! But!!! !n config/initializers/assets.rb I wrote:
**config.assets.compile = true**
All @import "file.css" in css-file work fine.
Whats the f*ck?!!! Sorry... Whats the BUG?!

**Comment by Sega100500** [comment]
@mkhairi
In Gemfile:
# Use SCSS for stylesheets
gem 'sass-rails', '~> 5.0'
In Gemfile.lock
sass (3.4.17)

How I can rollback?

**Comment by mkhairi** [comment]
@Sega100500 you can edit manually in Gemfile.lock then run ```bundle```

**Comment by seuros** [comment]
This discussion started to degenerated. 
@Sega100500 , if you can't keep your calm, i advice you to take some lesson of Yoga, there is a teacher in GTA 5.
```
gem 'sass-rails', '~> 5.0'
gem 'sass', '3.4.16'
```
bundle update sass

**Comment by Sega100500** [comment]
@mkhairi 
Yes. It fix problem

@seuros
Thanks!

**Comment by Sega100500** [comment]
https://github.com/sass/sass/issues/1810

**Comment by Sega100500** [comment]
But it bug sass only in Rails 4, in Rails 3 all work properly.
Whose is it still wrong?

**Comment by rafaelfranca** [comment]
Closing since it is a sass bug.

**Comment by Sega100500** [comment]
@rafaelfranca 
Then why is it manifested itself only in version RoR 4.2.4?
Coincidence?

**Comment by rafaelfranca** [comment]
Because when you update Rails it also update sass. I could reproduce this same problem with Rails 4.2.3 so it is not related to Rails 4.2.4

**Comment by Sega100500** [comment]
@rafaelfranca 
That is, you still got to reproduce the problem? You said that everything works.

**Comment by rafaelfranca** [comment]
Yeah, I could. My last commit on that application was reproducing the problem.

**Comment by chriseppstein** [comment]
Fixed in sass 3.4.18 which was just released.

**Comment by atussec** [comment]
this is not sass bug. his path are fixed path and not using asset pipeline.
for assets to be loaded in production it have to be define like this 
`@font-face {
    font-family: "enhance";
    src: asset-url("enhance.eot");
    src: asset-url("fonts/enhance.eot?#iefix") format("embedded-opentype"), asset-url("fonts/enhance.woff") format("woff"), asset-url("fonts/enhance.ttf") format("truetype"), asset-url("fonts/enhance.svg#enhance") format("svg");
    font-weight: normal;
    font-style: normal;
}`
